src/file-manager/fm-desktop-icon-view.c

```diff
@@ -708,11 +784,64 @@ action_change_background_callback (GtkAction *action,
 {
     g_assert (FM_DIRECTORY_VIEW (data));
 
-    caja_launch_application_from_command (gtk_widget_get_screen (GTK_WIDGET (data)),
-                                          _("Background"),
-                                          "mate-appearance-properties",
-                                          FALSE,
-                                          "--show-page=background", NULL);
+    if (GDK_IS_X11_DISPLAY (gdk_screen_get_display (gdk_screen_get_default())))
+        caja_launch_application_from_command (gtk_widget_get_screen (GTK_WIDGET (data)),
+                                              _("Background"),
+                                              "mate-appearance-properties",
+                                              FALSE,
+                                              "--show-page=background", NULL);
+    else
+    {
+        GtkWidget *dialog, *box;
+        char *filename;
+        GtkWidget *applybutton, *cancelbutton, *filebutton;
+        gint response;
+
+        filename = g_settings_get_string (background_settings, "picture-filename");
```

Comment:
Memory leak

Reply:
Plugged this at the end of the function (see below), and changed filename to the expected `gchar` so again we use` g_free()` when finished with it.